### PR TITLE
Push data to match expressions

### PR DESCRIPTION
**Pull request by michaelcfanning**
@eddynaka @jameswinkler @yongyan-gh

### CODE REVIEW

File: Src/Plugins/Security/SEC1001.SecurePlaintextSecrets.json [hunk 1]
```diff
@@ -3,8 +3,8 @@
   "Definitions": [
     {
       "SharedStrings": {
-        "BinaryFiles": "(?i)\\.(bmp|dll|exe|gif|jpg|jpeg|lock|png|psd|tar\\.gz|tif?|ttf|xcf|zip)$",
-        "SourceFiles": "(?i)\\.(azure|bat|c|cmd|config|cpp|cs|cscfg|definitions|dtsx|h|hxx|hpp|ini|java|jsx?|json|keys|kt|loadtest|m|md|php|properties|ps1|psm1|pubxml|py|resx|sample|sql|ste|swift|test|tsx?|txt|waz|xml)$"
+        "$BinaryFiles": "(?i)\\.(bmp|dll|exe|gif|jpg|jpeg|lock|png|psd|tar\\.gz|tif?|ttf|xcf|zip)$",
```
Comment (michaelcfanning) [inline Src/Plugins/Security/SEC1001.SecurePlaintextSecrets.json:6]: >$BinaryFiles [](start = 9, length = 12)

All shared strings are now denoted with a leading $ prefix, even at definition time.

File: Src/Plugins/Tests.Security/EndToEndTests.cs [hunk 2]
```diff
@@ -117,10 +130,18 @@ protected void RunAllTests()
 
             string testsDirectory = Path.Combine(ProductTestDataDirectory, @"Inputs\");
 
+            var inputFiles = new List<string>();
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/EndToEndTests.cs:133]: >inputFiles [](start = 16, length = 10)

This change was required to run all tests in the directory, to generate a comprehensive diff (we previously stopped after first file failure).

File: Src/Test.UnitTests.Sarif.PatternMatcher/FingerprintTests.cs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+using System.Reflection;
+using System.Text;
+
+using FluentAssertions;
+
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher
+{
+    public class FingerprintTests
```
Comment (michaelcfanning) [inline Src/Test.UnitTests.Sarif.PatternMatcher/FingerprintTests.cs:15]: >FingerprintTests [](start = 17, length = 16)

@eddynake, @jameswinkler @yongyan-gh, I'd like you to take a particular look at the patterns in these unit tests, to discuss further later.

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs [hunk 4]
```diff
@@ -122,36 +122,79 @@ internal static void PushInheritedData(SearchDefinitions definitions, Dictionary
...
+                                                             sharedStrings);
+
                 matchExpression.FileNameAllowRegex ??= definition.FileNameAllowRegex;
 
-                matchExpression.ContentsRegex = PushData(matchExpression.ContentsRegex, sharedStrings);
+                matchExpression.ContentsRegex = PushData(matchExpression.ContentsRegex,
+                                                         definition.SharedStrings,
+                                                         sharedStrings);
+
+                if (matchExpression.Level == 0)
```
Comment (eddynaka) [inline Src/Sarif.PatternMatcher/AnalyzeCommand.cs:151]: >0 [](start = 45, length = 1)

should we compare with FailureLevel.None, so people will know what is the  enum?

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs [hunk 5]
```diff
@@ -122,36 +122,79 @@ internal static void PushInheritedData(SearchDefinitions definitions, Dictionary
...
+                {
+                    matchExpression.Level = definition.Level;
+                }
+            }
+        }
+
+        private static string PushData(string text, params Dictionary<string, string>[] sharedStringsDictionaries)
+        {
+            if (text?.Contains("$") != true)
```
Comment (eddynaka) [inline Src/Sarif.PatternMatcher/AnalyzeCommand.cs:160]: >Contains [](start = 22, length = 8)

should we move to startswith? Why: that could be a real regex character.

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs [hunk 6]
```diff
@@ -122,36 +122,79 @@ internal static void PushInheritedData(SearchDefinitions definitions, Dictionary
...
+
+            return text;
         }
 
         internal static Dictionary<string, string> LoadSharedStrings(string sharedStringsFullPath, IFileSystem fileSystem)
         {
             var result = new Dictionary<string, string>();
 
-            foreach (string line in fileSystem.FileReadAllLines(sharedStringsFullPath))
+            foreach (string fileLine in fileSystem.FileReadAllLines(sharedStringsFullPath))
```
Comment (eddynaka) [inline Src/Sarif.PatternMatcher/AnalyzeCommand.cs:185]: >FileReadAllLines [](start = 51, length = 16)

nit: we could change from ReadAllLines to ReadLines. The first adds everything to memory the other uses ienumerable, so, one line is in memory.

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs [hunk 7]
```diff
@@ -122,36 +122,79 @@ internal static void PushInheritedData(SearchDefinitions definitions, Dictionary
...
+                if (string.IsNullOrEmpty(line) || line.StartsWith("#")) { continue; }
+
                 int index = line.IndexOf('=');
                 if (index == -1) { ThrowInvalidSharedStringsEntry(line); }
 
                 string key = line.Substring(0, index);
                 if (!key.StartsWith("$")) { ThrowInvalidSharedStringsEntry(line); }
 
-                result[key] = line.Substring(key.Length + "=".Length);
+                string value = line.Substring(key.Length + "=".Length);
```
Comment (eddynaka) [inline Src/Sarif.PatternMatcher/AnalyzeCommand.cs:196]: >string value [](start = 16, length = 12)

do we need to do this?

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs [hunk 8]
```diff
@@ -181,26 +224,11 @@ protected override ISet<Skimmer<AnalyzeContext>> CreateSkimmers(AnalyzeOptions o
...
-
-            return text;
-        }
-
         private static void ThrowInvalidSharedStringsEntry(string line)
         {
             throw new InvalidOperationException(
-                "Malformed shared strings entry. Every shared string should consist of a " +
-                "key name (prefixed with $) followed by an equals sign and the string value " +
+                $"Malformed shared strings entry. Every shared string should consist of a " +
```
Comment (eddynaka) [inline Src/Sarif.PatternMatcher/AnalyzeCommand.cs:230]: >$ [](start = 16, length = 1)

nit: remove if not needed.

**Review by eddynaka — APPROVED** [review]

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 5 above)
Comment (michaelcfanning) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:160]: Your suggestion won't allow shared strings to be embedded in the regex, e.g.,:

$HttpString$QueryString

Yes, $ can be a real regex character, but its presence (when not used for a shared string) will only provoke an unnecessary processing of the string (looking for variables to expand that don't actually exist).

---
In reply to: [555676556](https://github.com/microsoft/sarif-pattern-matcher/pull/101#discussion_r555676556) [](ancestors = 555676556)

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 8 above)
Comment (michaelcfanning) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:230]: This is for style, it makes all the string literals line up exactly. Does it bother you? :)

---
In reply to: [555679916](https://github.com/microsoft/sarif-pattern-matcher/pull/101#discussion_r555679916) [](ancestors = 555679916)

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 7 above)
Comment (michaelcfanning) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:196]: do what? declare a short-lived variable?

---
In reply to: [555679637](https://github.com/microsoft/sarif-pattern-matcher/pull/101#discussion_r555679637) [](ancestors = 555679637)

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 6 above)
Comment (michaelcfanning) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:185]: good suggestion.

---
In reply to: [555679261](https://github.com/microsoft/sarif-pattern-matcher/pull/101#discussion_r555679261) [](ancestors = 555679261)

**Pull request merged**